cf/data/data.py

```diff
@@ -3831,304 +3831,70 @@ def _binary_operation(self, other, method):
-                partition.update_inplace_from(partition_s)
+        try:
+            if method == "__eq__":
+                result = _numpy_isclose(dx0, dx1, rtol=rtol, atol=atol)
+            elif method == "__ne__":
+                result = ~_numpy_isclose(dx0, dx1, rtol=rtol, atol=atol)
+            elif inplace:
+                # Find non-in-place equivalent operator (remove 'i')
+                equiv_method = method[:2] + method[3:]
+                result = getattr(dx0, equiv_method)(dx1)
             else:
-                partition = partition_s
-
-            # --------------------------------------------------------
-            # Do the binary operation on this partition's data
-            # --------------------------------------------------------
-            try:
-                if method == "__eq__":  # and data0.Units.isreftime:
-                    array0 = _numpy_isclose(
-                        array0, array1, rtol=rtol, atol=atol
-                    )
-                elif method == "__ne__":
-                    array0 = ~_numpy_isclose(
-                        array0, array1, rtol=rtol, atol=atol
-                    )
-                else:
-                    array0 = getattr(array0, method)(array1)
-
-            except FloatingPointError as error:
-                # Floating point point errors have been trapped
-                if _mask_fpe[0]:
-                    # Redo the calculation ignoring the errors and
-                    # then set invalid numbers to missing data
-                    np.seterr(**_seterr_raise_to_ignore)
-                    array0 = getattr(array0, method)(array1)
-                    array0 = np.ma.masked_invalid(array0, copy=False)
-                    np.seterr(**_seterr)
-                else:
-                    # Raise the floating point error exception
-                    raise FloatingPointError(error)
-            except TypeError as error:
-                if inplace:
-                    raise TypeError(
-                        "Incompatible result data-type ({0!r}) for "
-                        "in-place {1!r} arithmetic".format(
-                            np.result_type(array0.dtype, array1.dtype).name,
-                            array0.dtype.name,
-                        )
+                result = getattr(dx0, method)(dx1)
+
+        except FloatingPointError as error:
+            # Floating point point errors have been trapped
+            if _mask_fpe[0]:
+                # Redo the calculation ignoring the errors and
+                # then set invalid numbers to missing data
+                np.seterr(**_seterr_raise_to_ignore)
+                result = getattr(dx0, method)(dx1)
+                result = np.ma.masked_invalid(dx0, copy=False)
+                np.seterr(**_seterr)
+            else:
+                # Raise the floating point error exception
+                raise FloatingPointError(error)
+        except TypeError as error:
```

Comment:
I think that this exception can go, too, as `dask` handles these cases OK:
```
>>> x = da.from_array([1, 2, 3])
>>> y = da.from_array(3.14)
>>> x += y
>>> x
dask.array<add, shape=(3,), dtype=float64, chunksize=(3,), chunktype=numpy.ndarray>
```

Reply:
Good spot, thanks. Nice, simplifies the logic further! Will get these catches removed.